ruby/phase-1.md

```diff
@@ -0,0 +1,598 @@
+  workflows: [MyWorkflow],
+  activities: [MyActivity]
+)
+worker.run
+```
+
+
+## Client
+
+The client is the primary point for interacting with workflows and async activities as well as
+managing namespaces. A client is created with a configuration object that specifies options shared
+between a worker and a client.
+
+```ruby
+config = Temporal::Configuration.new(url: 'localhost:7233')
+client = Temporal::Client.new(config)
+client.start_workflow(MyWorkflow, 'arg_1', kwarg_2: 'foo', workflow: {
+  namespace: 'my-namespace',
+  task_queue: 'my-queue'
+})
+```
+
+Notes:
+
+- Client is intended to be multi-tenant capable (no limit on a single namespace)
+- Configuration will provide defaults where possible (avoiding extensive upfront configuration)
+- Configuration will be used for other more advanced options such as logger, metrics, payload
+  converters, middleware, etc
+
+```ruby
+class Temporal::Client
+  # Creates a new instance of a Client
+  def self.new(config: Temporal::Configuration, **options) -> Temporal::Client
+
+  # Start the workflow and return handle for further interaction
+  # TODO: expand on :workflow
+  def start_workflow(
+    workflow, # workflow class or a name
+    *args, # any args to be passed as the input to the workflow
+    workflow: {}, # execution options
+    **kwargs # keyword arguments as inputs are also supported
+  ) -> Temporal::WorkflowHandle
+
+  # Start the workflow and block awaiting it's completion
+  def execute_workflow(
+    workflow, # workflow class or a name
+    *args, # any args to be passed as the input to the workflow
+    workflow: {}, # execution options
+    **kwargs # keyword arguments as inputs are also supported
+  ) -> Any
+
+  # Generate a handle object from a namespace, workflow_id and a run_id
+  def workflow_handle(namespace, workflow_id, run_id = nil) -> Temporal::WorkflowHandle
+end
+```
+
+Notes:
+
+- This is not a complete definition of the client
+- Both positional arguments and keyword arguments are supported, except for the reserved `options:`
+- Conventions similar to Python SDK are used here:
+  - `Client` is a general purpose interface, not limited workflow-only interactions
+  - Due to this method names are more descriptive than just `#start`, `#execute`, etc
+  - `Client#execute_workflow` is added for convenience
+
+Once a workflow is started all the interactions will be going through the handle object returned to
+the caller. The handle can of course be created from a workflow_id/run_id combination using the
+`Client#workflow_handle` method. Here the interface of the handle:
+
+```ruby
+class Temporal::WorkflowHandle
+  # Return the result of the workflow execution or raise if a workflow is still running after
+  #   waiting for for up to :timeout seconds.
+  # Passing timeout: nil will not wait if the working is still running
+  def result(timeout: Integer) -> Any
+
+  # Returns more information about a workflow
+  def describe() -> Temporal::WorkflowExecutionInfo
```

Comment:
This is somewhat inconsistent across existing Go, Java, and Ruby SDKs today.

Describe maps to a GRPC call whose response contains the execution info as well as three other fields ([source](https://github.com/temporalio/api/blob/4c2f6a281fa3fde8b0a24447de3e0d0f47d230b4/temporal/api/workflowservice/v1/request_response.proto#L557)), but they are often used interchangeably.

Go provides the entire payload as the proto generated class. The Java and Coinbase Ruby SDKs only provide the execution info portion of the payload. Java uses the proto classes and Ruby, its own wrapper classes.

Specific to this call, it would be good to clarify the behavior here to support the full set of outputs.

In general, it would be good to make a call on using the proto generated classes directly or wrappers for all APIs. Personally, I'd like to see wrapper classes for all of these, and not have proto generated classes be used as inputs or return values anywhere.

Reply:
Go is an outlier and cannot be changed.

I think what we want is for all [non-Go] SDKs' high-level clients to have a describe that returns a lang-friendly response that at least returns something like status at a high level but also provides access to the underlying gRPC response (e.g. a "raw" field). What other high-level fields are made available as friendly accessors to the underlying raw value I think is open for interpretation (`status` is a good candidate due to proto enums usually being worse than language-defined ones).

This already kinda exists in TypeScript, Python, and .Net and I wouldn't be surprised if we add to Java soon (https://github.com/temporalio/sdk-features/issues/17). In my mind, since we have so few high-level client calls anyways, not requiring the user to use/import proto classes at all for most common use cases is ideal.